*** mac/Emacs.app/Contents/Info.plist	2012-03-03 06:05:20 +0000
--- mac/Emacs.app/Contents/Info.plist	2012-03-06 03:31:31 +0000
***************
*** 71,76 ****
--- 71,78 ----
  	</array>
  	<key>NSAppleScriptEnabled</key>
  	<string>YES</string>
+ 	<key>NSPrincipalClass</key>
+ 	<string>EmacsApplication</string>
    	<key>NSServices</key>
  	<array>
  		<dict>

*** src/macappkit.h	2012-03-03 06:05:20 +0000
--- src/macappkit.h	2012-03-05 03:44:55 +0000
***************
*** 774,779 ****
--- 774,780 ----
  - (NSWindowAnimationBehavior)animationBehavior;
  - (void)setAnimationBehavior:(NSWindowAnimationBehavior)newAnimationBehavior;
  - (void)toggleFullScreen:(id)sender;
+ - (CGFloat)backingScaleFactor;
  @end
  #endif
  
***************
*** 850,854 ****
--- 851,859 ----
  + (void)runAnimationGroup:(void (^)(NSAnimationContext *context))changes
          completionHandler:(void (^)(void))completionHandler;
  @end
+ 
+ @interface CALayer (AvailableOn1070AndLater)
+ @property CGFloat contentsScale;
+ @end
  #endif
  #endif

*** src/macappkit.m	2012-03-03 06:05:20 +0000
--- src/macappkit.m	2012-03-05 03:44:55 +0000
***************
*** 2912,2917 ****
--- 2912,2918 ----
  #endif
    layer.bounds = CGRectMake (0, 0, [bitmap pixelsWide], [bitmap pixelsHigh]);
    layer.contents = (id) [bitmap CGImage];
+   layer.contentsScale = [transitionWindow backingScaleFactor];
    layer.contentsGravity = kCAGravityTopLeft;
    transitionContentView = [transitionWindow contentView];
    [transitionContentView setLayer:layer];



